tools/module_diag_hailcast.F90

```diff
@@ -0,0 +1,1852 @@
+! Remodeled as recommended for PR #164.
+!
+
+MODULE module_diag_hailcast
+    USE time_manager_mod, ONLY: time_type
+    use mpp_mod,          only: mpp_error, FATAL, input_nml_file, stdlog, mpp_pe, mpp_root_pe
+    use fms_mod,          only: check_nml_error
+    !use fv_mp_mod,        only: is_master
+
+    use constants_mod,    only: grav
+
+    logical :: do_hailcast = .false. !This controls whether hailcast is used
+
+    INTEGER :: id_hailcast_dhail
+    INTEGER :: id_hailcast_dhail1, id_hailcast_dhail2, id_hailcast_dhail3, &
+               id_hailcast_dhail4, id_hailcast_dhail5, id_hailcast_wdur,   &
+               id_hailcast_wup_mask
+    INTEGER :: id_hailcast_diam_mean, id_hailcast_diam_std
+    REAL, ALLOCATABLE :: hailcast_dhail1(:,:), hailcast_dhail2(:,:),    &
+                         hailcast_dhail3(:,:), hailcast_dhail4(:,:),    &
+                         hailcast_dhail5(:,:) !hailstone diameters (mm)
+    REAL, ALLOCATABLE :: hailcast_dhail1_max(:,:), hailcast_dhail2_max(:,:), &
+                         hailcast_dhail3_max(:,:), hailcast_dhail4_max(:,:), &
+                         hailcast_dhail5_max(:,:) !hailstone diameters (mm)
+    REAL, ALLOCATABLE :: hailcast_diam_mean(:,:), hailcast_diam_std(:,:) !mean and standard deviation of five hailstones (mm)
+    REAL, ALLOCATABLE :: hailcast_wdur(:,:), hailcast_wup_mask(:,:) !persistent arrays for updraft duration (s) and mask
+    real, allocatable :: hailcast_dhail_max(:,:)
+
+    integer :: kstt_hc, kend_hc
+    integer :: kstt_hc1,kstt_hc2,kstt_hc3,kstt_hc4,kstt_hc5
+    integer :: kend_hc1,kend_hc2,kend_hc3,kend_hc4,kend_hc5
+    integer :: kstt_hcd, kend_hcd, kstt_hcm, kend_hcm
+
+    PRIVATE :: INTERPP, INTERP, TERMINL, VAPORCLOSE, MASSAGR, HEATBUD, BREAKUP, MELT
+    PRIVATE :: hailstone_driver, hailcast_diagnostic_driver
+CONTAINS
+
+    SUBROUTINE hailcast_init(file_name, axes, Time, isdo,iedo,jsdo,jedo,nlevs, missing_value, istatus)
+
+        use diag_manager_mod,   only: register_diag_field
+
+        CHARACTER(LEN=64), INTENT(IN)    :: file_name
+        integer,           intent(in)    :: axes(4)
+        type(time_type),   intent(in)    :: Time
+        INTEGER,           INTENT(IN)    :: isdo,iedo,jsdo,jedo
+        INTEGER,           INTENT(INOUT) :: nlevs
+        real,              INTENT(IN)    :: missing_value
+        INTEGER,           INTENT(OUT)   :: istatus
+
+        namelist /fv_diagnostics_nml/ do_hailcast
+        integer :: ios, ierr
+        integer :: unit, f_unit
+
+        !namelist file for hailcast
+#ifdef INTERNAL_FILE_NML
+        ! Read Main namelist
+        read (input_nml_file,fv_diagnostics_nml,iostat=ios)
+        ierr = check_nml_error(ios,'fv_diagnostics_nml')
+#else
+        f_unit=open_namelist_file()
+        rewind (f_unit)
+        ! Read Main namelist
+        read (f_unit,fv_diagnostics_nml,iostat=ios)
+        ierr = check_nml_error(ios,'fv_diagnostics_nml')
+        call close_file(f_unit)
+#endif
```

Comment:
We are no longer supporting the monolithic open_namelist option in favor of the INTERNAL_FILE_NML logic.  Please remove the #ifdef-#endif and include only the read from input_nml_file logic.